### ISSUE TITLE
Number of constraints in Blake2s (and optimization)

### ISSUE THREAD

**Issue by sanchopansa** [akosba/xjsnark#17]
I've been working on an implementation of Blake2s in xjsnark (mps file is attached).

The circuit works as expected, but I'm getting a total of 21912 constraints for it. Looking at the [Zcash protocol specification](https://github.com/zcash/zips/blob/master/protocol/protocol.pdf) (page 133) the number of constraints for Blake2s is estimated at 21262. I realize that the estimation in the Zcash spec is assuming a manually constructed R1CS and to me the benefit of coding this in a DSL outweighs the difference of less than 700 constraints, but I am still curious if you know why there is this discrepancy. Maybe something in my xjsnark implementation can be optimized to match the manual implementation?

As a side note, I tried applying the same optimization technique that is used in the SHA-256 example:
```
Config.multivariateExpressionMinimization = true; 
 
// Num of threads could be set higher based on the machine specs. Setting it too high will lead to faster timeouts 
Config.arithOptimizerNumThreads = 4; 
Config.arithOptimizerTimeoutPerProblemMilliSec = 20000;
```
However, I left it running for about 2 hours on my machine and it was still ongoing (SHA-256 optimization takes less than a minute I think). I'm also curious to understand why that is the case, if you have an idea.

Thank you!

[blake.mps.zip](https://github.com/akosba/xjsnark/files/2838752/blake.mps.zip)

**Comment by akosba** [comment]
> The circuit works as expected, but I'm getting a total of 21912 constraints for it. Looking at the [Zcash protocol specification](https://github.com/zcash/zips/blob/master/protocol/protocol.pdf) (page 133) the number of constraints for Blake2s is estimated at 21262. I realize that the estimation in the Zcash spec is assuming a manually constructed R1CS and to me the benefit of coding this in a DSL outweighs the difference of less than 700 constraints, but I am still curious if you know why there is this discrepancy. Maybe something in my xjsnark implementation can be optimized to match the manual implementation?

The reason is that one of the opcodes that jsnark outputs (pack) can be implemented in 0 constraints, but it's currently implemented using [1 constraint](https://github.com/akosba/jsnark/blob/f44c8ca50cf5f1f02396475974e93295064e9927/JsnarkCircuitBuilder/src/circuit/operations/primitive/PackBasicOp.java#L68). As mentioned, this will require modifying this part of the [interface](https://github.com/akosba/libsnark/blob/5aec0646a26c7dc41b70195b744659e67abf3554/libsnark/jsnark_interface/CircuitReader.cpp#L536). In the Blake2s circuit, there are 636 pack operations in the output circuit, which could explain the difference between 21912 and 21262. Packing a number of bits can be implemented as a weighted linear combination without adding any constraints, but I had it initially done the other way, as I noticed that introducing one new variable via a new constraint might reduce a bit of overhead for some kinds of circuits where the packed result (or a linear combination of many variables in general) is used several times, but I have not revisited that. I had a plan to change the implementation of pack and linear combinations to support both options, and select the better representation based on the circuit shape, but I thought the savings will not be much, so did not give that a priority.  I will change the interface soon to support both possibilities for pack.

> However, I left it running for about 2 hours on my machine and it was still ongoing (SHA-256 optimization takes less than a minute I think). I'm also curious to understand why that is the case, if you have an idea.

Thanks for noticing that. There was a performance issue in the optimizer. It was not limiting the size of the expressions as it's supposed to in some special cases that appear in the structure of this circuit. Since this circuit is xor-heavy, unrolling the expressions using + and * operations made them much larger, requiring more time. SHA256 had a different structure that did not have this problem. I pushed a fix (note that this won't help reduce the number of constraints for this circuit, but will just make the optimizer run smoothly this time).
